Fix gravity strength, dash behavior, add ability for directional walk/dash and improve PlatformPhysics

henriiquecampos opened the pull request:
Close #9 #12 #13 #15 #16 and #17

henriiquecampos committed:
Fix typo and add ability to enable or disable air-dash

henriiquecampos committed:
Update dash to prevent vertical movement when dashing

henriiquecampos committed:
Update gravity and jump strength and

henriiquecampos committed:
Add signals to communicate deplete and replenish of available jumps

Connect said signals to Dash to prevent the character from dashing if it doesn't have available jumps, i.e. character cannot dash after double jump

henriiquecampos commented:
Hey @salvob41 would you mind testing this iteration? I wish to know if it makes sense, especially tracking the signals through the inspector. Do you think it is too "hidden"?

nitaku commented:
Hey, great work! :) I was just about to make a basic platformer in Godot myself, along with @salvob41

I have some comments from a design standpoint.

I feel that dashing and double jumping could be more interesting if they have less mechanical overlap. For example, you could completely disable falling while dashing. This would make dashing a better option when the character has to cross the gaps between the first two platforms. This is what happens now if you attempt to do that:

![dash](https://user-images.githubusercontent.com/1604569/57767169-098a9980-7709-11e9-889d-708fae2f8abe.gif)

It ultimately depends on the type of game you want to make though. :)

henriiquecampos commented:
> I feel that dashing and double jumping could be more interesting if they have less mechanical overlap. For example, you could completely disable falling while dashing.

That is exactly what I tried to achieve in [`line 12 on Dash.gd`](https://github.com/pigdevstudio/kitchen-tales/pull/14/files#diff-64835658073a6ad47427805e626c4bd7R12).

@nitaku did you test this iteration on the [`development` branch](https://github.com/pigdevstudio/kitchen-tales/tree/development)? It seems to be working :thinking: 

![002 - Dash stops gravity](https://user-images.githubusercontent.com/10171059/57772435-c3730d00-76eb-11e9-91b9-5d14db5aae9a.gif)

Check it out and tell me if it works, if it doesn't maybe gravity is still being applied in some extend and I must fix it, because this is one of the behaviors I want to achieve with dash as well. Thanks for the heads-up :grin:

henriiquecampos committed:
Add ability to perform infinite jumps on Jump by setting max jumps to -1

henriiquecampos committed:
Fix Ground Geometry2D shape to snap to the 8px grid

nitaku commented:
Whoops, you are absolutely right! I thought I switched to `development` but I didn't. ^_^' It works as expected.

salvob41 commented on source/actors/Jump.gd:
```diff
@@ -8,18 +10,23 @@ onready var _available_jumps = max_jumps
 func apply():
-	if _available_jumps > 0:
-		plaform_actor.velocity.y = -strength
+	if _available_jumps > 0 or max_jumps == -1:
```
here it is safer adding a `<=-1` instead of `==` even if this condition should never happened. 

But I think all these hardcoded numbers make the code hardly understandable. I would go with maybe nested condition and enum that describes behaviours

salvob41 replied:
scratch the first part, this is definitely a case of enum

salvob41 commented on source/actors/PlatformActor.tscn:
```diff
@@ -39,14 +39,16 @@ script = ExtResource( 7 )
 distance = 400.0
 
-[connection signal="dash_pressed" from="ActionBus" to="Dash" method="set_physics_process" binds= [ true ]]
+[connection signal="available_jumps_depleted" from="Jump" to="Dash" method="set_air_dash" binds= [ false ]]
```
I think that two behaviours should not be directly connected, that's why we are trying to use a "Manager" right? 
Once Jump "replenished"  or "depleted" their actions, should send a signal to the Manager who will act accordingly - imagine something else happens when your jumps are depleted, like you can make a crash on the floor pressing the button down ( like the one you can do with Mario, don't know if it has a name), if this is the case you should connect multiple signal to different components and that is not elegant

henriiquecampos replied:
Yeah I didn't implement the manager yet. I am still trying to foresee if using the Inspector's interface can be enough. Since the manager will only tell what triggers what, it must not have conditionals etc...it would be basically an *ad hoc* implementation to make these signals connections explicit

Still wrapping my head around this manager

salvob41 reviewed (changes requested):
Hey Pigdev, thanks for asking me. I would gladly review the code but here's a few things that I don't think goes well with all the behaviour-manager system, and I would like to discuss with you guys.
1. Connecting signals between behaviours will tidy up strictly the two components, and I don't think we want that (I put more detail in the comment review)
2. Personally I prefer connecting signals from code, so I can see which signal is connected to which node directly from the code instead of the scene. Plus If I create a new element with those behaviours, or delete a node and then create a new one, I don't have to do all the things by hand over and over
3?. ActionBus looks like a manager of Inputs. It sounds like a Handler more than a Manager. I am seeing it more like the parent of all behaviours (Jump, Dash) and instead of sending signals I would, maybe, calling directly the node's method - what do you think about it @nitaku ?
4. I think all the node behaviour should have a `enabled` flag. Imagine there is a potion, or a magic spell that make the Character unable to jump. It is actually able to jump, normally, but if that "component" is disabled, it could not. In this way you won't change the tree and everything would work normally

salvob41 commented:
Playing around with the actual implementation I found myself able to do this:

![kitchen-tales-dash](https://user-images.githubusercontent.com/6097015/57781407-6cd8f380-772a-11e9-83ab-76cdff56ca21.gif)

Plus, I actually liked the inertia (now disappeared, on `master` branch is still possible) that caused the character to fly faster if the dash is activated just before a jump. Like the celeste [dash&jump](https://www.youtube.com/watch?v=Gds9V6vAsQw) 


P.S.: I vote for directional dash (like celeste)

henriiquecampos commented:
Thanks a lot for the insights @salvob41 continuing the discussion...
>     1. Connecting signals between behaviours will tidy up strictly the two components, and I don't think we want that (I put more detail in the comment review)

I think `signals` are the most loose way to communicate events or pass data between objects...as directly calling the functions would hard couple them, so I think their communication should use their signals. For instance, imagine reusing each of these components for other games, something I tried with the `Jump` for instance. If I was making a Flappy Bird copy, what would I need to do in order to make the character flap? Currently I would just take the PlatformActor, add the Gravity, ActionBus and the Jump and connect their signals.

Anything you know that can provide a more decoupled/loose approach for that reusability?
 
>     2. Personally I prefer connecting signals from code, so I can see which signal is connected to which node directly from the code instead of the scene. Plus If I create a new element with those behaviours, or delete a node and then create a new one, I don't have to do all the things by hand over and over

Yeah I think it is more explicit to see these connections via code...that's something I am considering. Tho there is a downside. Making the connections via the Inspector will throw the responsibility of managing and updating changes in the scene hierarchy to the engine, since it provides automatic updates to the signal connections' NodePaths...plus a UI to manage said connections...That's some advantages that I am still considering if losing them is worthy.

>        3?. ActionBus looks like a manager of Inputs. It sounds like a Handler more than a Manager. I am seeing it more like the parent of all behaviours (Jump, Dash) and instead of sending signals I would, maybe, calling directly the node's method - what do you think about it @nitaku ?

`I am seeing it more like the parent of all behaviours` I thought that as well! Definitely, but I kept this way due to a design/abstraction question: Who makes the jump/dash/walk, the **PlatformActor** or the **ActionBus**?

I see the ActionBus more as an Input element, that is part of Player's character, which I can take off for PlatformActors that are not controlled by the Player. Let's say, create an `AIBrain` that would handle decisions or emulate Actions for jump/dash/walk on other `PlatformActors` that are not players, e.g. a boss character. So the ActionBus is not a part of the general PlatformActor, but the behaviors are. That's why I kept them as siblings.

>     3. I think all the node behaviour should have a `enabled` flag. Imagine there is a potion, or a magic spell that make the Character unable to jump. It is actually able to jump, normally, but if that "component" is disabled, it could not. In this way you won't change the tree and everything would work normally

This is achieved through the `set_physics_process` method you can see that everywhere. E.g. all the `ActionBus` does is to set said process `enabled` or `disabled` depending on the InputEvent.

Tho it might be a good approach to add a property that would intermediate that to make it more explicit. Something on the `PlatformPhysics` like:

```GDScript
class_name PlaformPhysics
extends Node

onready var platform_actor = get_node(platform_actor_path)

export (NodePath) var platform_actor_path = ".."
export (bool) var enabled = true setget set_enabled

func set_enabled(value):
    enabled = value
    set_physics_process(enabled)
```

Wachu think?

henriiquecampos commented:
> Playing around with the actual implementation I found myself able to do this:
> 
> ![kitchen-tales-dash](https://user-images.githubusercontent.com/6097015/57781407-6cd8f380-772a-11e9-83ab-76cdff56ca21.gif)
> 
> Plus, I actually liked the inertia (now disappeared, on `master` branch is still possible) that caused the character to fly faster if the dash is activated just before a jump. Like the celeste [dash&jump](https://www.youtube.com/watch?v=Gds9V6vAsQw)
> 
> P.S.: I vote for directional dash (like celeste)

Any reproducible steps for that? :scream:

henriiquecampos committed:
Fix #15

henriiquecampos commented:
> Playing around with the actual implementation I found myself able to do this:
> 
> ![kitchen-tales-dash](https://user-images.githubusercontent.com/6097015/57781407-6cd8f380-772a-11e9-83ab-76cdff56ca21.gif)

Ohh thanks a loooooooooooot. I just fixed that!

> Plus, I actually liked the inertia (now disappeared, on `master` branch is still possible) that caused the character to fly faster if the dash is activated just before a jump. Like the celeste [dash&jump](https://www.youtube.com/watch?v=Gds9V6vAsQw)

Yeah I am going to re-add it, just trying to figure out a way to do so. My main reference is Megaman X tho, but the idea is the same: if you start dashing in the ground and jump, the jump will keep the dash horizontal speed.

> P.S.: I vote for directional dash (like celeste)

Wachu you mean by `direction dash`? I didn't play Celeste yet

salvob41 commented:
The first point was more related to the "Manager" I thought you already started implemented that. 

> Anything you know that can provide a more decoupled/loose approach for that reusability?

Signals are actually the best thing to do that. So let's stick with that. It makes me wonder if sending signals would have performance issues, but I actually don't think so, hearing from Godot devs.

> So the ActionBus is not a part of the general PlatformActor, but the behaviors are. That's why I kept them as siblings. 

sounds fair. :D 

> Wachu think?

I had something like this in mind. In this way the code is more elegant as well :) approved.

> Wachu you mean by direction dash? I didn't play Celeste yet 

You should play Celeste! It will give you lots of idea about platformers in general. Apart from the fact that it is a GREAT game, The gameplay and mechanics feel soooo good. 
Talking about "directional dash", here there is a [wonderful explanation about the Celeste dash](https://www.rockpapershotgun.com/2018/04/03/celeste-dash/). As you can see if you press the UP button and the dash, it will dash upwards. Same applies for all other 8 directions. It feels good in a platform

henriiquecampos commented:
> sending signals would have performance issues, but I actually don't think so, hearing from Godot devs.

With the implementation I used I don't think it might be an issue, the signals are only emitted when the events happen, e.g. they are not constantly being emitted like 60 signals emitting in `_process`

henriiquecampos committed:
Add apply() to PlatformPhysics to allow proper polymorphism

henriiquecampos committed:
Add ability to enable or disable PlatformPhysics using a config bool

henriiquecampos committed:
Add ability to pass Vectors to Walk movement_direction

henriiquecampos committed:
Update Dash to have independent conditions check

Make the code more readable and allow for decoupled conditions, might be interesting in the future

henriiquecampos commented:
>You should play Celeste! It will give you lots of idea about platformers in general. Apart from the fact that it is a GREAT game, The gameplay and mechanics feel soooo good.
Talking about "directional dash", here there is a wonderful explanation about the Celeste dash. As you can see if you press the UP button and the dash, it will dash upwards. Same applies for all other 8 directions. It feels good in a platform

Well this is not on the design of Kitchen Tales, but [I added the ability](https://github.com/pigdevstudio/kitchen-tales/pull/14/commits/e93484f65ffc15d76e9f8d78571c5e1806ca8f02#diff-f77a6670018bb28bdee4e0310a3cb5d3R4) to do that by allowing to set the `Walk.movement_direction` using a `Vector2` and `Dash` now uses both axis.

You can have a `handle_direction` on the `ActionBus` to check if Up or Down buttons are pressed and change the `Vector2` accordingly

Check [ActionBus.handle_walk's direction property](https://github.com/pigdevstudio/kitchen-tales/pull/14/commits/dfd163e85272389c63940b98ef424c3ebef2410a#diff-0e394490c2b1fca2c8c0426caad2e36cR48). With that you can have some checks for vertical inputs and threat the vector properly.

henriiquecampos committed:
Update ActionBus.handle_walk to threat the input direction instead of filter it

henriiquecampos commented:
Alright I think it is all good, thank you so much @salvob41 and @nitaku ! I hope this PR is useful for you guys too :heart:

salvob41 commented:
Thank you for the opportunity. Tag me in the next PR that I'll see what I can do :)
